#####Text for shiny app

introduction <- 'Welcome to the analytical module of bacLIFE, a Shiny app that allows you to easily interact with the results generated by previous modules of our tool (you can read more about each module in [reference to the paper]). Here, you will be able to explore your data exhaustively, going from the gene level to the biosynthetic gene clusters (BGCs) present. To use this Shiny app, you need to execute the complete bacLIFE pipeline ([github page]). An overview of bacLIFE is shown below:'



bacLIFE_app <- 'The bacLIFE app is a comprehensive tool designed to analyze the distribution of genes and biosynthetic gene clusters across genomes. It is composed of five sections, each with specific functions and features aimed at providing users with a visual representation of the vast amounts of data generated by the analysis. Let’s look at each section of the app: '

exploration <- 'This first section enables the user to perform Principal Coordinate Analysis (PCoA) based on dice dissimilarity and dendrogram visualization of the gene level and Biosynthetic Gene (BGC) level absence/presence matrices, both levels being divided into different tabs in the app. Users can annotate the plots with the different columns present in the metadata file: mapping_file.txt. The gene level matrix can also be aggregated by a database ID such as COG, KEGG, DBCAN, and PROKKA. BGCs can also be visualized in a heatmap.'

clustering <- "The clustering section offers different methods for clustering (knn, hdbscan, cut dendrogram) in order to generate new bacterial groups and metadata. Clustering can be performed on the two principal components (visualized in the previous section) of the chosen matrix by knn or hdbscan, and cutting the dendrogram created with the chosen matrix”. Once the clustering is done, the user can visualize the distribution of the predicted lifestyles in each cluster made in the tab 'Lifestyle Preservation' "


corepananalysis <- 'Pancore analysis, provides an overview of the pancore-gene clusters or pancore-BGC clusters defined by the original metadata or the new bacterial groups new metadata generated in clustering. The COG hierarchy functional annotation of the gene clusters present in each genome is shown in the "Pan-Genome" tab. Users can group genomes by metadata categories to view the variance in functionality and number of genes. The "Core-Genome" section presents the functions of the core-gene clusters of the complete dataset as well as the core-gene clusters of metadata groups and their overlap. The number of unique genes per sample can also be extracted from the subsection “Unique Genes”, and a brief overview of the COG categories and their description can be observed in the last subsection: “About”'

statisticalanalysis <- 'This section generates different plots to visualize the gene level and BGC level statistics between the previously established bacterial groups, generated in the Clustering section. The statistics can be done at the gene level (CD-HIT, COG, KEGG, DBCAN, PROKKA) and at the BGC level (per Gene Cluster Family - GCF). Fisher test in combination with basic relative presence statistics are used to study the distribution of the gene clusters among the metadata.” Since these are paired tests, two groups must be selected by the user for the comparison. Once selected, the button “Go” can be pressed and the user can move onto the “Statistics” subsection, where the plots will be shown. The user also has the option to look for specific queries in the new mapping file (shown in the main page of this section) and even download those selected queries or the full table.'

geneclusterdistribution <- "Enables users to check the distribution of gene clusters of interest by introducing the id of the gene cluster. The Shiny app generates a PCoA indicating the presence/absence of the gene in the different bacteria."

downloadfasta <- "Users can download the sequences in fasta format of a gene cluster or database id of interest of the complete dataset or specific bacteria"
